gcp/src/main/java/org/apache/iceberg/gcp/biglake/BigLakeClient.java

```diff
@@ -0,0 +1,270 @@
+import java.util.Map;
+import java.util.function.Supplier;
+import org.apache.iceberg.exceptions.AlreadyExistsException;
+import org.apache.iceberg.exceptions.NoSuchNamespaceException;
+import org.apache.iceberg.exceptions.NoSuchTableException;
+import org.apache.iceberg.exceptions.NotAuthorizedException;
+
+/** A client of Google BigLake service. */
+final class BigLakeClient implements Closeable {
+
+  private final MetastoreServiceClient stub;
+
+  /**
+   * Constructs a client of Google BigLake Service.
+   *
+   * @param settings BigLake service settings
+   */
+  BigLakeClient(MetastoreServiceSettings settings) throws IOException {
+    this.stub = MetastoreServiceClient.create(settings);
+  }
+
+  /**
+   * Constructs a client of Google BigLake Service.
+   *
+   * @param biglakeEndpoint BigLake service gRPC endpoint, e.g., "biglake.googleapis.com:443"
+   */
+  BigLakeClient(String biglakeEndpoint) throws IOException {
+    this(MetastoreServiceSettings.newBuilder().setEndpoint(biglakeEndpoint).build());
+  }
+
+  public Catalog createCatalog(CatalogName name, Catalog catalog) {
+    return convertException(
+        () -> {
+          try {
+            return stub.createCatalog(
+                CreateCatalogRequest.newBuilder()
+                    .setParent(LocationName.of(name.getProject(), name.getLocation()).toString())
+                    .setCatalogId(name.getCatalog())
+                    .setCatalog(catalog)
+                    .build());
+          } catch (com.google.api.gax.rpc.AlreadyExistsException e) {
+            throw new AlreadyExistsException(e, "Catalog already exists: %s", name.getCatalog());
+          }
+        },
+        name.getCatalog());
+  }
+
+  public Catalog catalog(CatalogName name) {
+    try {
+      return stub.getCatalog(GetCatalogRequest.newBuilder().setName(name.toString()).build());
+    } catch (PermissionDeniedException e) {
+      throw new NoSuchNamespaceException(
+          e, "Catalog does not exist: %s (or permission denied)", name.getCatalog());
+    }
+  }
+
+  public void deleteCatalog(CatalogName name) {
+    try {
+      stub.deleteCatalog(DeleteCatalogRequest.newBuilder().setName(name.toString()).build());
+    } catch (PermissionDeniedException e) {
+      throw new NoSuchNamespaceException(
+          e, "Catalog does not exist: %s (or permission denied)", name.getCatalog());
+    }
+  }
+
+  public Database createDatabase(DatabaseName name, Database db) {
+    return convertException(
+        () -> {
+          try {
+            return stub.createDatabase(
+                CreateDatabaseRequest.newBuilder()
+                    .setParent(
+                        CatalogName.of(name.getProject(), name.getLocation(), name.getCatalog())
+                            .toString())
+                    .setDatabaseId(name.getDatabase())
+                    .setDatabase(db)
+                    .build());
+          } catch (com.google.api.gax.rpc.AlreadyExistsException e) {
```

Comment:
Shouldn't this also handle `PermissionDeniedException` since that may be thrown by any of the calls? Here it should probably be translated to Iceberg's `ForbiddenException` or `NotAuthorizedException`. Or is this an appropriate place to return `NoSuchNamespaceException` to indicate that the catalog doesn't exist?

Reply:
The wrapper convertException will handle PermissionDeniedException and convert it to Iceberg's NotAuthorizedException. It is extracted to a wrapper for reducing boilerplate code.

Reply:
Yes, it is a good point. We should check whether it failed due to parent not found or permission denied. It needs parsing the error message. I added a TODO to do it in a follow-up PR, to avoid adding more new code to this PR which has been around for long.